beacon-chain/core/peerdas/helpers.go

```diff
@@ -0,0 +1,248 @@
+package peerdas
+
+import (
+	"encoding/binary"
+
+	cKzg4844 "github.com/ethereum/c-kzg-4844/bindings/go"
+	"github.com/holiman/uint256"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	errors "github.com/pkg/errors"
+	fieldparams "github.com/prysmaticlabs/prysm/v5/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v5/config/params"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/interfaces"
+	"github.com/prysmaticlabs/prysm/v5/crypto/hash"
+	"github.com/prysmaticlabs/prysm/v5/encoding/bytesutil"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+)
+
+const (
+	// Number of field elements per extended blob
+	fieldElementsPerExtBlob = 2 * cKzg4844.FieldElementsPerBlob
+
+	// Bytes per cell
+	bytesPerCell = cKzg4844.FieldElementsPerCell * cKzg4844.BytesPerFieldElement
+
+	// Number of cells in the extended matrix
+	extendedMatrixSize = fieldparams.MaxBlobsPerBlock * cKzg4844.CellsPerExtBlob
+)
+
+type (
+	extendedMatrix []cKzg4844.Cell
+
+	cellCoordinate struct {
+		blobIndex uint64
+		cellID    uint64
+	}
+)
+
+var (
+	errCustodySubnetCountTooLarge          = errors.New("custody subnet count larger than data column sidecar subnet count")
+	errCellNotFound                        = errors.New("cell not found (should never happen)")
+	errCurveOrder                          = errors.New("could not set bls curve order as big int")
+	errBlsFieldElementNil                  = errors.New("bls field element is nil")
+	errBlsFieldElementBiggerThanCurveOrder = errors.New("bls field element higher than curve order")
+	errBlsFieldElementDoesNotFit           = errors.New("bls field element does not fit in BytesPerFieldElement")
+)
+
+// https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/das-core.md#helper-functions
+func custodyColumns(nodeId enode.ID, custodySubnetCount uint64) (map[uint64]bool, error) {
+	dataColumnSidecarSubnetCount := params.BeaconConfig().DataColumnSidecarSubnetCount
+
+	// Compute the custodied subnets.
+	subnetIds, err := CustodyColumnSubnets(nodeId, custodySubnetCount)
+	if err != nil {
+		return nil, errors.Wrap(err, "custody subnets")
+	}
+
+	columnsPerSubnet := cKzg4844.CellsPerExtBlob / dataColumnSidecarSubnetCount
+
+	// Knowing the subnet ID and the number of columns per subnet, select all the columns the node should custody.
+	// Columns belonging to the same subnet are contiguous.
+	columnIndices := make(map[uint64]bool, custodySubnetCount*columnsPerSubnet)
+	for i := uint64(0); i < columnsPerSubnet; i++ {
+		for subnetId := range subnetIds {
+			columnIndex := dataColumnSidecarSubnetCount*i + subnetId
+			columnIndices[columnIndex] = true
+		}
+	}
+
+	return columnIndices, nil
+}
+
+func CustodyColumnSubnets(nodeId enode.ID, custodySubnetCount uint64) (map[uint64]bool, error) {
```

Comment:
```suggestion
func CustodyColumnSubnets(nodeId enode.ID) (map[uint64]bool, error) {
```
I do not think the custody count should be an argument, it is not dynamic but a value determined by the specification

Reply:
Actually the spec defines [get_custody_columns](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/das-core.md#get_custody_columns) with the custody count as an argument:

```python
def get_custody_columns(node_id: NodeID, custody_subnet_count: uint64) -> Sequence[ColumnIndex]:
    assert custody_subnet_count <= DATA_COLUMN_SIDECAR_SUBNET_COUNT

    subnet_ids = []
    i = 0
    while len(subnet_ids) < custody_subnet_count:
        subnet_id = (
            bytes_to_uint64(hash(uint_to_bytes(uint64(node_id + i)))[0:8])
            % DATA_COLUMN_SIDECAR_SUBNET_COUNT
        )
        if subnet_id not in subnet_ids:
            subnet_ids.append(subnet_id)
        i += 1
    assert len(subnet_ids) == len(set(subnet_ids))

    columns_per_subnet = NUMBER_OF_COLUMNS // DATA_COLUMN_SIDECAR_SUBNET_COUNT
    return [
        ColumnIndex(DATA_COLUMN_SIDECAR_SUBNET_COUNT * i + subnet_id)
        for i in range(columns_per_subnet)
        for subnet_id in subnet_ids
    ]
```

The custody subnet count has an impact of the number of custody column subnet count.

>  it is not dynamic but a value determined by the specification

My understanding is the specification determines [CUSTODY_REQUIREMENT](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/das-core.md#custody-setting)  (Minimum number of subnets an honest node custodies and serves samples from).
```go
params.BeaconConfig().CustodyRequirement // in Prysm
```

`CUSTODY_REQUIREMENT` is a **minimum** custody subnet count, but it is possible for a node to custody more. (And in the extreme case, a "supernode" can custody all the subnets.)

In my opinion, both for spec compliance and possibility of custodying more than the minimum required, it's better to keep the custody count as an argument.